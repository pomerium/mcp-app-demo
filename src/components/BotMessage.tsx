import { Bot, Copy, Download } from 'lucide-react'
import { toast } from 'sonner'
import { useState } from 'react'
import { MessageAvatar } from './MessageAvatar'
import type { AssistantStreamEvent } from '@/hooks/useStreamingChat'
import type { AnnotatedFile } from '@/lib/utils/code-interpreter'
import { cn, formatTimestamp } from '@/lib/utils'
import { MarkdownContent } from '@/components/MarkdownContent'
import { copyToClipboard } from '@/lib/utils/clipboard'
import {
  createAnnotatedFileUrl,
  isImageFile,
  replaceSandboxUrls,
} from '@/lib/utils/code-interpreter'
import { ContentRenderer } from './ContentRenderer'
import type { Content, UIAction } from '@/types/mcp'

export interface Message extends Omit<AssistantStreamEvent, 'type'> {
  timestamp: string
  status: string
}

export interface BotMessageProps {
  message: Message
  fileAnnotations?: Array<AnnotatedFile>
  onUIAction?: (action: UIAction) => void
}

export function BotMessage({
  message,
  fileAnnotations = [],
  onUIAction,
}: BotMessageProps) {
  const [imageErrors, setImageErrors] = useState<Set<string>>(new Set())
  const processedContent = replaceSandboxUrls(message.content, fileAnnotations)

  const handleCopy = async () => {
    const copied = await copyToClipboard(processedContent)

    if (copied) {
      toast.success('Copied message to clipboard')
    }
  }

  const handleImageError = (fileId: string) => {
    setImageErrors((prev) => new Set(prev).add(fileId))
  }

  const fileLikeAnnotations = fileAnnotations.filter(
    (annotation) => annotation.type === 'container_file_citation',
  )
  const imageFiles = fileLikeAnnotations.filter((annotation) =>
    isImageFile(annotation.filename),
  )
  const otherFiles = fileLikeAnnotations.filter(
    (annotation) => !isImageFile(annotation.filename),
  )

  return (
    <div
      className={cn(
        'flex w-full max-w-full gap-2 py-2 animate-in fade-in',
        'justify-start',
      )}
    >
      <MessageAvatar icon={<Bot className="h-5 w-5" />} variant="default" />
      <div
        className={cn(
          'group grid gap-1 space-y-1 items-start',
          'w-full sm:w-[85%] md:w-[75%] lg:w-[65%]',
        )}
      >
        <div
          className={cn(
            'rounded-2xl px-4 py-2 text-sm w-full bg-gray-100 text-gray-900 dark:bg-gray-800 dark:text-gray-100',
          )}
        >
          {/* Render MCP content items if present */}
          {message.mcpContent && message.mcpContent.length > 0 && (
            <div className="space-y-2 mb-2">
              {message.mcpContent.map((item: any, index: number) => {
                if (item && typeof item === 'object' && 'type' in item) {
                  const content = item as Content
                  return (
                    <ContentRenderer
                      key={index}
                      content={content}
                      fileAnnotations={fileAnnotations}
                      onUIAction={onUIAction}
                    />
                  )
                }
                return null
              })}
            </div>
          )}

          {/* Render markdown content if present */}
          {processedContent && (
            <div data-raw-markdown={processedContent}>
              <MarkdownContent
                content={processedContent}
                fileAnnotations={fileAnnotations}
              />
            </div>
          )}

          {imageFiles.length > 0 && (
            <div className="mt-4 space-y-3">
              {imageFiles.map((annotation) => {
                const hasError = imageErrors.has(annotation.file_id)
                const imageUrl = createAnnotatedFileUrl(annotation)

                const getAltText = (filename: string) => {
                  const name = filename.replace(/\.[^/.]+$/, '')
                  return `Generated visualization: ${name.replace(/_/g, ' ')}`
                }

                return (
                  <div key={annotation.file_id} className="relative group">
                    {hasError ? (
                      <div
                        className="flex items-center justify-center p-4 bg-gray-50 dark:bg-gray-900 rounded border border-dashed border-gray-300 dark:border-gray-600"
                        role="alert"
                      >
                        <div className="text-center">
                          <p className="text-sm text-gray-500 dark:text-gray-400">
                            Failed to load image
                          </p>
                          <a
                            href={createAnnotatedFileUrl(annotation)}
                            download={annotation.filename}
                            className="mt-2 text-xs text-blue-600 dark:text-blue-400 hover:underline"
                          >
                            Download file instead
                          </a>
                        </div>
                      </div>
                    ) : (
                      <>
                        <img
                          src={imageUrl}
                          alt={getAltText(annotation.filename)}
                          className="max-w-full h-auto rounded border border-gray-200 dark:border-gray-700 cursor-pointer"
                          onError={() => handleImageError(annotation.file_id)}
                          onClick={() => window.open(imageUrl, '_blank')}
                        />
                        <a
                          href={createAnnotatedFileUrl(annotation)}
                          download={annotation.filename}
                          className="absolute top-2 right-2 p-2 bg-white/80 dark:bg-gray-800/80 backdrop-blur-sm rounded-full shadow-sm opacity-0 group-hover:opacity-100 transition-opacity hover:bg-white dark:hover:bg-gray-800"
                          title="Download image"
                        >
                          <Download className="h-4 w-4 text-gray-600 dark:text-gray-300" />
                        </a>
                      </>
                    )}
                  </div>
                )
              })}
            </div>
          )}

          {otherFiles.length > 0 && (
            <div className="mt-4 space-y-2">
              <p className="text-xs text-gray-500 dark:text-gray-400 font-medium">
                Attachments:
              </p>
              <ul className="space-y-1" aria-label="File attachments">
                {otherFiles.map((annotation) => (
                  <li
                    key={annotation.file_id}
                    className="flex items-center justify-between p-2 bg-gray-50 dark:bg-gray-900 rounded border border-gray-200 dark:border-gray-700"
                  >
                    <span className="text-sm text-gray-700 dark:text-gray-300 truncate">
                      {annotation.filename}
                    </span>
                    <a
                      href={createAnnotatedFileUrl(annotation)}
                      download={annotation.filename}
                      className="flex items-center gap-1 px-2 py-1 text-xs text-blue-600 dark:text-blue-400 hover:bg-blue-50 dark:hover:bg-blue-900/20 rounded"
                    >
                      <Download className="h-3 w-3" />
                      Download
                      <span className="sr-only"> {annotation.filename}</span>
                    </a>
                  </li>
                ))}
              </ul>
            </div>
          )}
        </div>
        <div className="flex gap-1 items-center text-xs text-gray-500 dark:text-gray-400 space-x-1">
          <time dateTime={message.timestamp}>
            {formatTimestamp(new Date(message.timestamp))}
          </time>
          <button
            type="button"
            className="p-1 rounded hover:bg-gray-200 dark:hover:bg-gray-700 focus:outline-none focus:ring-2 focus:ring-primary"
            onClick={handleCopy}
          >
            <Copy className="h-3 w-3" aria-hidden="true" />
            <span className="sr-only">Copy message to clipboard</span>
          </button>
        </div>
      </div>
    </div>
  )
}
