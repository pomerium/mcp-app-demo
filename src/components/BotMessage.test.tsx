// @vitest-environment jsdom
import { render } from '@testing-library/react'
import { describe, it, expect } from 'vitest'
import { BotMessage } from './BotMessage'
import type { Message } from '../mcp/client'

const markdownTable = `| Year | Total Sales        |
|------|-------------------|
| 2012 | $18,823,201.72    |
| 2013 | $38,633,120.01    |
| 2014 | $38,870,148.13    |
| 2015 | $41,423,456.72    |
| 2016 | $40,568,672.36    |
| 2017 | $40,209,904.23    |
| 2018 | $38,326,623.43    |
| 2019 | $38,516,963.86    |
| 2020 | $38,862,436.79    |
| 2021 | $41,355,549.74    |
| 2022 | $39,742,066.18    |
| 2023 | $33,054,490.00    |
\nHere is a summary table of total sales from every year. If you need additional detail, just let me know!`

describe('BotMessage', () => {
  it('renders a markdown table correctly', () => {
    const message: Message = {
      id: 'msg-1',
      content: markdownTable,
      sender: 'agent',
      timestamp: new Date('2025-07-06T12:00:00Z'),
      status: 'sent',
    }
    const { container } = render(<BotMessage message={message} />)
    // Table header and a few years should be present
    expect(container).toMatchSnapshot()
  })
})
