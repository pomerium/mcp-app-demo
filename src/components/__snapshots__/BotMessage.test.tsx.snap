// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BotMessage > renders a markdown table correctly 1`] = `
<div>
  <div
    class="flex w-full max-w-full gap-2 py-2 animate-in fade-in justify-start"
  >
    <div
      class="flex h-8 w-8 shrink-0 select-none items-center justify-center rounded-md bg-gray-100 text-gray-900 dark:bg-gray-800 dark:text-gray-100"
    >
      <svg
        class="lucide lucide-bot h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 8V4H8"
        />
        <rect
          height="12"
          rx="2"
          width="16"
          x="4"
          y="8"
        />
        <path
          d="M2 14h2"
        />
        <path
          d="M20 14h2"
        />
        <path
          d="M15 13v2"
        />
        <path
          d="M9 13v2"
        />
      </svg>
    </div>
    <div
      class="group grid gap-1 space-y-1 items-start w-full sm:w-[85%] md:w-[75%] lg:w-[65%]"
    >
      <div
        class="rounded-2xl px-4 py-2 text-sm w-full bg-gray-100 text-gray-900 dark:bg-gray-800 dark:text-gray-100"
      >
        <div
          data-raw-markdown="| Year | Total Sales        |
|------|-------------------|
| 2012 | $18,823,201.72    |
| 2013 | $38,633,120.01    |
| 2014 | $38,870,148.13    |
| 2015 | $41,423,456.72    |
| 2016 | $40,568,672.36    |
| 2017 | $40,209,904.23    |
| 2018 | $38,326,623.43    |
| 2019 | $38,516,963.86    |
| 2020 | $38,862,436.79    |
| 2021 | $41,355,549.74    |
| 2022 | $39,742,066.18    |
| 2023 | $33,054,490.00    |

Here is a summary table of total sales from every year. If you need additional detail, just let me know!"
        >
          <div
            class="prose prose-sm dark:prose-invert max-w-none break-words break-all whitespace-pre-wrap"
          >
            <div
              class="prose prose-sm dark:prose-invert max-w-none break-words [&>*+*]:mt-4 [&>h1+*]:mt-3 [&>h2+*]:mt-3 [&>h3+*]:mt-3 [&>p+ul]:mt-3 [&>p+ol]:mt-3"
            >
              <div
                class="overflow-x-auto my-4"
              >
                <table>
                  <thead>
                    <tr>
                      <th>
                        Year
                      </th>
                      <th>
                        Total Sales
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr>
                      <td>
                        2012
                      </td>
                      <td>
                        $18,823,201.72
                      </td>
                    </tr>
                    <tr>
                      <td>
                        2013
                      </td>
                      <td>
                        $38,633,120.01
                      </td>
                    </tr>
                    <tr>
                      <td>
                        2014
                      </td>
                      <td>
                        $38,870,148.13
                      </td>
                    </tr>
                    <tr>
                      <td>
                        2015
                      </td>
                      <td>
                        $41,423,456.72
                      </td>
                    </tr>
                    <tr>
                      <td>
                        2016
                      </td>
                      <td>
                        $40,568,672.36
                      </td>
                    </tr>
                    <tr>
                      <td>
                        2017
                      </td>
                      <td>
                        $40,209,904.23
                      </td>
                    </tr>
                    <tr>
                      <td>
                        2018
                      </td>
                      <td>
                        $38,326,623.43
                      </td>
                    </tr>
                    <tr>
                      <td>
                        2019
                      </td>
                      <td>
                        $38,516,963.86
                      </td>
                    </tr>
                    <tr>
                      <td>
                        2020
                      </td>
                      <td>
                        $38,862,436.79
                      </td>
                    </tr>
                    <tr>
                      <td>
                        2021
                      </td>
                      <td>
                        $41,355,549.74
                      </td>
                    </tr>
                    <tr>
                      <td>
                        2022
                      </td>
                      <td>
                        $39,742,066.18
                      </td>
                    </tr>
                    <tr>
                      <td>
                        2023
                      </td>
                      <td>
                        $33,054,490.00
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
              

              <p>
                Here is a summary table of total sales from every year. If you need additional detail, just let me know!
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex gap-1 items-center text-xs text-gray-500 dark:text-gray-400 space-x-1"
      >
        <time
          datetime="2025-07-06T12:00:00.000Z"
        >
          8:00 AM
        </time>
        <button
          class="p-1 rounded hover:bg-gray-200 dark:hover:bg-gray-700 focus:outline-none focus:ring-2 focus:ring-primary"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-copy h-3 w-3"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="14"
              rx="2"
              ry="2"
              width="14"
              x="8"
              y="8"
            />
            <path
              d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
            />
          </svg>
          <span
            class="sr-only"
          >
            Copy message to clipboard
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
